File: src/13.dynamicProgramming/01.arrayDP/08.650.ts
/**
 * 动态规划其实是运筹学的一种最优化方法
 * 一般形式就是求最值
 * 求最值核心问题: 需要穷举
 * 动态规划的穷举特点: 存在重叠子问题, 一定会具备最优子结构
 * 明确 baseCase -> 明确状态 -> 明确选择 -> 定义 dp 数组
 */

export function minSteps(n: number): number {
    const dp = new Array(n + 1).fill(0).map(() => {
        return new Array(n + 1).fill(n);
    });

    dp[1][1] = 0;

    for (let i = 1; i <= n; i++) {
        let minNum = dp[i][1];

        for (let j = 1; j <= i; j++) {
            if (i - j >= 1) {
                dp[i][j] = Math.min(dp[i][j], dp[i - j][j] + 1);
                minNum = Math.min(minNum, dp[i][j]);
            }
            if (i == j) {
                dp[i][j] = minNum + 1;
            }
        }
    }

    return dp[n][n] - 1;
}

File: src/13.dynamicProgramming/01.arrayDP/08.650.go
package arraydp

func MinSteps(n int) int {
	// n 个 A
	// k 个 A 需要的次数
	// f(k) =
	// const dp = ne
	dp := make([][]int, n+1)

	for i := 0; i <= n; i++ {
		dp[i] = make([]int, n+1)
		for j := 0; j <= n; j++ {
			dp[i][j] = n
		}
	}

	dp[1][1] = 0

	for i := 1; i <= n; i++ {
		minNum := dp[i][1]

		for j := 1; j <= i; j++ {
			if i-j >= 1 {
				val := dp[i-j][j] + 1
				if dp[i][j] > val {
					dp[i][j] = val
				}

				if minNum > dp[i][j] {
					minNum = dp[i][j]
				}
			}

			if i == j {
				dp[i][j] = minNum + 1
			}
		}
	}

	return dp[n][n] - 1
}
